stan/math/rev/functor/reduce_sum.hpp

```diff
@@ -453,35 +259,19 @@ struct reduce_sum_impl<ReduceFunction, M, T, Arg1, Arg2, Arg3, Arg4, var> {
         tbb::blocked_range<std::size_t>(0, num_jobs, grainsize), worker);
 #endif
 
-    std::vector<std::size_t> num_terms_arg(5, 0);
-
-    num_terms_arg[0] = op_vmapped.num_elements();
-    num_terms_arg[1] = worker.op_arg1_.num_elements();
-    num_terms_arg[2] = worker.op_arg2_.num_elements();
-    num_terms_arg[3] = worker.op_arg3_.num_elements();
-    num_terms_arg[4] = worker.op_arg4_.num_elements();
-
-    const std::size_t num_terms = sum(num_terms_arg);
-
     vari** varis
         = ChainableStack::instance_->memalloc_.alloc_array<vari*>(num_terms);
     double* partials
         = ChainableStack::instance_->memalloc_.alloc_array<double>(num_terms);
 
-    std::size_t idx = 0;
+    save_varis(varis, args...);
 
-    op_vmapped.build(&varis[idx], &partials[idx]);
-    idx += num_terms_arg[0];
-    worker.op_arg1_.build(&varis[idx], &partials[idx]);
-    idx += num_terms_arg[1];
-    worker.op_arg2_.build(&varis[idx], &partials[idx]);
-    idx += num_terms_arg[2];
-    worker.op_arg3_.build(&varis[idx], &partials[idx]);
-    idx += num_terms_arg[3];
-    worker.op_arg4_.build(&varis[idx], &partials[idx]);
+    for (size_t i = 0; i < num_terms; ++i) {
+      partials[i] = worker.args_adjoints_(i);
+    }
```

Comment:
Could do a parallel_for here?

Reply:
That's not worthwhile, really. The bulk of the work is getting the gradients of the sub-expressions. This is a mere copying in memory which probably could be done faster with a `std::copy` call.